File: src/main/java/org/lanternpowered/server/item/firework/LanternFireworkEffectBuilder.java
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.item.firework;

import static com.google.common.base.Preconditions.checkNotNull;

import org.spongepowered.api.item.FireworkEffect;
import org.spongepowered.api.item.FireworkShape;
import org.spongepowered.api.item.FireworkShapes;
import org.spongepowered.api.util.Color;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class LanternFireworkEffectBuilder implements FireworkEffect.Builder {

    private boolean trail;
    private boolean flicker;
    private List<Color> colors;
    private List<Color> fades;
    private FireworkShape shape;

    public LanternFireworkEffectBuilder() {
        reset();
    }

    @Override
    public LanternFireworkEffectBuilder trail(boolean trail) {
        this.trail = trail;
        return this;
    }

    @Override
    public LanternFireworkEffectBuilder flicker(boolean flicker) {
        this.flicker = flicker;
        return this;
    }

    @Override
    public LanternFireworkEffectBuilder color(Color color) {
        checkNotNull(color);
        this.colors.add(color);
        return this;
    }

    @Override
    public LanternFireworkEffectBuilder colors(Color... colors) {
        checkNotNull(colors);
        Collections.addAll(this.colors, colors);
        return this;
    }

    @Override
    public LanternFireworkEffectBuilder colors(Iterable<Color> colors) {
        checkNotNull(colors);
        for (Color color : colors) {
            this.colors.add(color);
        }
        return this;
    }

    @Override
    public LanternFireworkEffectBuilder fade(Color color) {
        checkNotNull(color);
        this.fades.add(color);
        return this;
    }

    @Override
    public LanternFireworkEffectBuilder fades(Color... colors) {
        checkNotNull(colors);
        Collections.addAll(this.fades, colors);
        return this;
    }

    @Override
    public LanternFireworkEffectBuilder fades(Iterable<Color> colors) {
        checkNotNull(colors);
        for (Color color : colors) {
            this.fades.add(color);
        }
        return this;
    }

    @Override
    public LanternFireworkEffectBuilder shape(FireworkShape shape) {
        this.shape = checkNotNull(shape);
        return this;
    }

    @Override
    public FireworkEffect build() {
        return new LanternFireworkEffect(this.flicker, this.trail, this.colors, this.fades, this.shape);
    }

    @Override
    public FireworkEffect.Builder from(FireworkEffect value) {
        return trail(value.hasTrail())
                .colors(value.getColors())
                .fades(value.getFadeColors())
                .shape(value.getShape())
                .flicker(value.flickers());
    }

    @Override
    public LanternFireworkEffectBuilder reset() {
        this.trail = false;
        this.flicker = false;
        this.colors = new ArrayList<>();
        this.fades = new ArrayList<>();
        this.shape = FireworkShapes.BALL;
        return this;
    }
}

File: src/main/kotlin/org/lanternpowered/server/effect/firework/LanternFireworkEffectBuilder.kt
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.effect.firework

import org.lanternpowered.api.effect.firework.FireworkEffect
import org.lanternpowered.api.effect.firework.FireworkEffectBuilder
import org.lanternpowered.api.effect.firework.FireworkShape
import org.lanternpowered.api.effect.firework.FireworkShapes
import org.lanternpowered.api.ext.*
import org.lanternpowered.api.util.Color
import java.util.ArrayList

class LanternFireworkEffectBuilder : FireworkEffectBuilder {

    private var trail: Boolean = false
    private var flicker: Boolean = false
    private val colors: MutableList<Color> = ArrayList()
    private val fades: MutableList<Color> = ArrayList()
    private var shape: FireworkShape? = null

    init {
        reset()
    }

    override fun trail(trail: Boolean): FireworkEffectBuilder = apply { this.trail = trail }
    override fun flicker(flicker: Boolean): FireworkEffectBuilder = apply { this.flicker = flicker }
    override fun shape(shape: FireworkShape): FireworkEffectBuilder = apply { this.shape = shape }
    override fun color(color: Color): FireworkEffectBuilder = apply { this.colors.add(color) }
    override fun colors(vararg colors: Color): FireworkEffectBuilder = apply { this.colors.addAll(colors) }
    override fun colors(colors: Iterable<Color>): FireworkEffectBuilder = apply { this.colors.addAll(colors) }
    override fun fade(color: Color): FireworkEffectBuilder = apply { this.fades.add(color) }
    override fun fades(vararg colors: Color): FireworkEffectBuilder = apply { this.fades.addAll(colors) }
    override fun fades(colors: Iterable<Color>): FireworkEffectBuilder = apply { this.fades.addAll(colors) }

    override fun build(): FireworkEffect {
        return LanternFireworkEffect(this.flicker, this.trail, this.colors.toImmutableList(), this.fades.toImmutableList(), this.shape!!)
    }

    override fun from(value: FireworkEffect): FireworkEffectBuilder = apply {
        trail(value.trail)
        colors(value.colors)
        fades(value.fadeColors)
        shape(value.shape)
        flicker(value.flickers)
    }

    override fun reset(): FireworkEffectBuilder = apply {
        this.trail = false
        this.flicker = false
        this.colors.clear()
        this.fades.clear()
        this.shape = FireworkShapes.BALL
    }
}
